(maint) Removes open_uri_redirections

mhashizume opened the pull request:
Prior to Ruby 2.4, the OpenURI module did not natively support HTTP => HTTPS redirections, necessitating a third-party gem, open_uri_redirections.

However, redirection functionality was added in Ruby 2.4:

https://github.com/ruby/ruby/commit/393ecc9f10016828a3713438ae216dc39b965364

And the `Kernel#open` method was deprecated in Ruby 2.7 and removed entirely in Ruby 3.0 in favor of the `URI#open` method:

https://bugs.ruby-lang.org/issues/15893

This commit removes the open_uri_redirections gem from the WebHelpers module and gemspec and replaces `Kernel#open` with `URI#open`.

mhashizume commented:
I need to figure out how to mock this in tests

ekohl reviewed (commented):
:+1: for removing a dependency and relying on stdlib more.

bastelfreak commented on .github/workflows/test.yml:
```diff
@@ -26,17 +26,17 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - ruby: "2.4"
```
dropping ruby 2.4 would be a breaking change. I want to discuss dropping 2.4/2.5/2.5 support tomorrow at cfgmgmtcamp.

mhashizume replied:
Sounds good. `URI#open` is weirdly a private method in 2.4 but we need it in >= 3.0, I'll find a way around it.

mhashizume commented:
- I updated the PR summary
- I removed an earlier commit that removed Ruby 2.4 from tests
- This PR is failing Rubocop tests because rubocop-performance has updated since I opened it and an updated cop is failing. I have [another PR open](https://github.com/voxpupuli/beaker/pull/1765) that will address these issues, and I will rebase this PR once the rubocop-specific PR has been merged.

ekohl commented on lib/beaker/dsl/helpers/web_helpers.rb:
```diff
@@ -117,6 +116,16 @@ def fetch_http_dir(url, dst_dir)
+        # Kernel#open approach in this code can be removed and this
+        # functionality can be reincorporated into fetch_http_file
+        private
+        def conditional_open(src)
```
I'd personally name it `compat_open` or even `uri_open_compat`.

ekohl commented on spec/beaker/dsl/helpers/web_helpers_spec.rb:
```diff
@@ -37,16 +37,25 @@ def fetch_allows
         create_files([concat_path])
         allow(logger).to receive(:notify)
-        allow(subject).to receive(:open)
+        # This can be simplified down to one case once we drop Ruby 2.4 support
```
How so? It's `conditional_open` called, regardless of Ruby version?

mhashizume commented:
@ekohl I saw your message on the Vox mailing list about moving Beaker to >= Ruby 2.7.

What were you thinking in terms of timeline? I can greatly simplify this PR if even Ruby 2.4 support is not an issue (and we won't have to revisit it later).

ekohl commented:
Soonish? I think @bastelfreak suggested to:
* Merge https://github.com/voxpupuli/beaker/pull/1765
* Update https://github.com/voxpupuli/beaker/pull/1759
* Release last version in the Beaker 4 series
* Start Beaker 5

So IMHO we can start this pretty quickly, though I'd like to note that I personally won't have much time to work on a major cleanup in the next 2 weeks or so.

bastelfreak commented:
I merged https://github.com/voxpupuli/beaker/pull/1759 and did the minor release. IMO we can now drop Ruby 2.4/2.5/2.6 + bump the ruby version in .rubocop.yml. I recommend the following gem versions (tested it in json-schema the past couple of days):

```
group :tests do
  gem 'rubocop', '~> 1.45.0'
  gem 'rubocop-rspec', '~> 2.18.1'
  gem 'rubocop-rake', '~> 0.6.0'
  gem 'rubocop-performance', '~> 1.16.0'
end
```

I will propose such a PR, when that's merged, #1764 can be rebased?

mhashizume commented:
Sounds good @bastelfreak

bastelfreak commented:
I created https://github.com/voxpupuli/beaker/pull/1767

bastelfreak commented:
@mhashizume can you please rebase? :)
